### PR TITLE
Map voting is now ticket-based

### PR DESCRIPTION
**Pull request by moxian**
## What Does This PR Do
Switched map vote counting to ticket (aka. urn) mode. Details at #17869

TL;DR:
each voting person casts a vote into the bag, then we draw one vote from the bag at random, and declare that to be the vote result.
Alternative way of visualizing that: Each vote candidate gets weighted by the amount of votes cast for them, then a random one is chosen. So an candidate with twice as many votes behind it is twice as likely to be chosen.

## Why It's Good For The Game
See #17869
TL;DR: promotes map variety, while being democratic.

## Images of changes
20220721-050716-dreamseeker

Wording suggestions are welcome (either here or on discord - both work).

One thing i'm a bit torn about is displaying the winning ticket holder in the message:
- on one hand i really like that it gives excellent feedback about how the system works, that it is *fair*, and that your vote *truly* counts no matter how many are already cast
- on the other hand, i slightly worry that it reveals that person's vote, which might subject them to bullying for voting "wrong".

I'm open to removing that part if that would be preferable.
Or maybe going halfway, and tying it to anon mode in prefs. So that instead of merely anonymizing the ckey, it stops even showing the player mob.

## Changelog
:cl:
tweak: Map vote counting is changed to be ticket-based (cast all votes into the bag; draw out a single random one - declare it as the winer)
/:cl:

stealth-change: effectively reverts #17838 and shows the vote count while the map vote is in progress again. There is no reason to be herdy anymore - each individual vote truly counts, no matter how small of a minority you are.

### CODE REVIEW

File: SQL/paradise_schema.sql [hunk 1]
```diff
@@ -351,6 +351,7 @@ CREATE TABLE `karma_totals` (
 ) ENGINE=InnoDB AUTO_INCREMENT=25715 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
 /*!40101 SET character_set_client = @saved_cs_client */;
 
+DROP TABLE IF EXISTS `karma_purchases`;
```
Comment (AffectedArc07) [inline SQL/paradise_schema.sql:354]: Something something atomicity.

File: code/modules/vote/vote_datum.dm [hunk 2]
```diff
@@ -59,21 +60,32 @@
...
+
+	// And log them to DB
+	if(!is_custom)
+		for(var/res in results)
+			SSblackbox.record_feedback("nested tally", "votes", results[res], list(vote_type_text, res), ignore_seal = TRUE)
+
+	// Here goes the actual winner selection code
+	switch(vote_counting)
+		if(VOTE_COUNTING_MAJORITY)
 			if(!length(voted))
```
Comment (AffectedArc07) [inline code/modules/vote/vote_datum.dm:85]: If we check this twice should we just move it to the common area?

File: code/modules/vote/vote_presets.dm [hunk 3]
```diff
@@ -17,6 +17,8 @@
 /datum/vote/map
 	question = "Map Vote"
 	vote_type_text = "map"
+	vote_counting = VOTE_COUNTING_TICKET
+	show_counts = TRUE
```
Comment (AffectedArc07) [inline code/modules/vote/vote_presets.dm:21]: I don't like that you've dressed this up as a stealth change, and the counts are still relevant in ticket voting as it still increases the weight. 

The votes will decide though (hue).

File: SQL/paradise_schema.sql (hunk 1 above)
Comment (moxian) [reply SQL/paradise_schema.sql]: Right!.. I'll move it to a separate PR.

File: code/modules/vote/vote_datum.dm (hunk 2 above)
Comment (moxian) [reply code/modules/vote/vote_datum.dm:85]: No because it's feasible to have a `VOTE_COUNTING_1984` in which votes don't actually matter, and lack of votes would still result in the win of The Correct Candidate.

File: code/modules/vote/vote_presets.dm (hunk 3 above)
Comment (moxian) [reply code/modules/vote/vote_presets.dm:21]: You are no longer incentivised to stick to the majority. So you no longer need to know who the majority is to do the right decision. So displaying the votes doesn't sabotage the vote one way or another.

In the `VOTE_COUNTING_MAJORITY`, for example, if you want "anything except Box", then you *have to* vote for whichever non-box map is currently in the lead, else your vote is completely wasted.
And this is not the problem at all with `VOTE_COUNTING_TICKET`

File: code/modules/vote/vote_datum.dm [hunk 4]
```diff
@@ -106,15 +114,31 @@
...
-			return null
+		if(VOTE_COUNTING_TICKET)
+			if(!length(voted))
+				to_chat(world, "<span class='interface'>No votes were cast. Do you all hate democracy?!</span>") // shame them
+				return null
+
+			var/lucky_winner_ck = pick(voted)
+			var/lucky_winner_vote = voted[lucky_winner_ck]
+			var/lucky_mob = get_mob_by_ckey(lucky_winner_ck)
+			lucky_winner_ck = safe_get_ckey(lucky_mob)  // anonymize, if requested
```
Comment (AffectedArc07) [inline code/modules/vote/vote_datum.dm:128]: Don't include the mob here, even if anonymised. This will lead to abuse in OOC for "why did you pick that" 

Just have it say reaching into the bag and the answer is xyz

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

File: code/modules/vote/vote_datum.dm (hunk 4 above)
Comment (moxian) [reply code/modules/vote/vote_datum.dm:128]: (I'm not yet sure how hard it would be to implement, but assuming it's not too hacky) what do you think about anonymizing, yet not completely removing the mob mention?
As in: "...vote cast by a certain medical/security/cargo/service/science department employee" / "a certain ghost" / "a certain Black Terror Spider" / "a certain Blob Overmind" / "a certain Ian" / ...

Comment (AffectedArc07) [reply code/modules/vote/vote_datum.dm:128]: Nah, OOC will still become a mess of "fuck that [mobname]", which will cause issues for simplemobs given:
- Theyre numbered in most cases 
- Mobs like Ian are singletons

Comment (moxian) [reply code/modules/vote/vote_datum.dm:128]: Numbers can be very easily stripped via `initial(mob.name)` (although honestly, who the hell remembers the mob numbers).

And even for singleton mobs like Ian, the OOC salt would only extend to the current [end of] round, since there's no link between the mob and the player's ckey or "main" IC name. So I see no opportunity for metagrudging.
(And people unhappy with the vote result would OOC salt anyway).

That said, I'm open to total removal of the mob mention. I just want to make sure that it is a somewhat real (and not a purely hypothetical) problem first.

**Comment by AffectedArc07** [comment]
Design team as voted against unfortunately. Shame.

I do have plans to end the only-box trend though.
